dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/data/upload/DataUploadRunnable.kt

```diff
@@ -33,53 +30,44 @@ internal class DataUploadRunnable(
     private val contextProvider: ContextProvider,
     private val networkInfoProvider: NetworkInfoProvider,
     private val systemInfoProvider: SystemInfoProvider,
-    uploadConfiguration: DataUploadConfiguration,
+    internal val uploadSchedulerStrategy: UploadSchedulerStrategy,
+    internal val maxBatchesPerJob: Int,
     private val internalLogger: InternalLogger
 ) : UploadRunnable {
 
-    internal var currentDelayIntervalMs = uploadConfiguration.defaultDelayMs
-    internal val minDelayMs = uploadConfiguration.minDelayMs
-    internal val maxDelayMs = uploadConfiguration.maxDelayMs
-    internal val maxBatchesPerJob = uploadConfiguration.maxBatchesPerUploadJob
-
     //  region Runnable
 
     @WorkerThread
     override fun run() {
+        var uploadAttempts = 0
+        var lastBatchUploadStatus: UploadStatus? = null
         if (isNetworkAvailable() && isSystemReady()) {
             val context = contextProvider.context
             var batchConsumerAvailableAttempts = maxBatchesPerJob
-            var lastBatchUploadStatus: UploadStatus?
             do {
                 batchConsumerAvailableAttempts--
                 lastBatchUploadStatus = handleNextBatch(context)
-            } while (batchConsumerAvailableAttempts > 0 &&
-                lastBatchUploadStatus is UploadStatus.Success
+                if (lastBatchUploadStatus != null) {
+                    uploadAttempts++
+                }
+            } while (
+                batchConsumerAvailableAttempts > 0 && lastBatchUploadStatus is UploadStatus.Success
             )
-            if (lastBatchUploadStatus != null) {
-                handleBatchConsumingJobFrequency(lastBatchUploadStatus)
-            } else {
-                // there was no batch left or there was a problem reading the next batch
-                // in the storage so we increase the interval
-                increaseInterval(lastBatchUploadStatus)
-            }
         }
 
-        scheduleNextUpload()
+        val delayMs = uploadSchedulerStrategy.getMsDelayUntilNextUpload(
+            featureName,
+            uploadAttempts,
+            lastBatchUploadStatus?.code,
+            lastBatchUploadStatus?.throwable
```

Comment:
What if we simply pass `lastBatchUploadStatus` as a whole here? Does it have a downside?

Reply:
I'm not keen of making the whole `UploadStatus` public. WDYT?